tinyformat: refactor: increase compile-time checks and don't throw for tfm::format_error

stickies-v opened the pull request:
Avoid unexpected run-time crashes from string formatting by enforcing compile-time format string checks for most* `tfm::format` and `strprintf` usage, and by returning an error string instead of throwing for a run-time tfm::format_error so callsites don't need to implement this error handling.

This PR should introduce no behaviour change.
The main changes are:
- remove the `const std::string&` `tfm::format` overload since it's not necessary. Usage of this overload is removed by one of:
  - replacing string concatenation in the format string with just an extra parameter
  - using the `bilingual_str` overload
  - using the `tfm::format_raw` functions (only for tinyformat implementation or tests)
- rename the non-compile-time-validated `tfm::format` overloads to `tfm::format_raw`, so existing callsites by default use the safer `util::ConstevalFormatString` `tfm::format` overloads. Callsites that for some reason don't pass the compile-time checks (such as in `bitcoin-cli.cpp`) can use `tfm::format_raw`.
- make `tfm::format_error` internal to `tinyformat` and just return error strings instead, so callsites don't need to implement error handling to prevent crashing. 

*The `bilingual_str format(const bilingual_str& fmt, const Args&... args)` is not yet compile-time checked, which means the `lint-format-strings.py` test can't be removed yet.

 Follow-up on https://github.com/bitcoin/bitcoin/pull/30889#discussion_r1766633746, an alternative to #15926.

l0rinc commented on src/util/string.h:
```diff
@@ -235,11 +235,27 @@ template <typename T1, size_t PREFIX_LEN>
+std::string try_format(util::ConstevalFormatString<sizeof...(Args)> fmt, const Args&... args)
 {
-    return format(fmt.fmt, args...);
+    std::string log_msg;
+    try {
+        log_msg = tfm::format(fmt.fmt, args...);
+    } catch (tinyformat::format_error& fmterr) {
+        /* Original format string will have newline so don't add one here */
+        log_msg = "Error \"" + std::string{fmterr.what()} + "\" while formatting log message: \"" + fmt.fmt + "\"";
+    }
+    return log_msg;
```
Since we have simple, atomic instructions in both branches after which we immediately return, we might as well return directly:
```suggestion
    try {
        return tfm::format(fmt.fmt, args...);
    } catch (tinyformat::format_error& fmterr) {
        /* Original format string will have newline so don't add one here */
        return "Error \""s + fmterr.what() + "\" while formatting log message: \"" + fmt.fmt + "\"";
    }
```

Note: I've used `using namespace std::string_literals` to simplify `"Error \""s + fmterr.what()` concat here.

stickies-v replied:
Taken (minus string literals), thanks!

l0rinc commented on src/netbase.cpp:
```diff
@@ -559,12 +559,8 @@ std::function<std::unique_ptr<Sock>(int, int, int)> CreateSock = CreateSockOS;
-        LogDebug(BCLog::NET, "%s\n", error_message);
-    }
+    auto log_level{manual_connection ? BCLog::Level::Info : BCLog::Level::Debug};
+    LogPrintLevel(BCLog::NET, log_level, fmt, args...);
```
Is the `LogFlags::NONE` -> `BCLog::NET` category in case of `manual_connection` a behavior change?

l0rinc commented on src/test/util_string_tests.cpp:
```diff
@@ -83,4 +84,12 @@ BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
 
+BOOST_AUTO_TEST_CASE(tinyformat_try_format)
+{
+    // ensure invalid format strings don't throw at run-time
```
comment seems redundant here

stickies-v replied:
I don't agree it's redundant:
- it's not obvious from the tests that we're not just checking error messages, but that we're ensuring no errors are thrown, which is the difference with `tfm::format()`
- additional tests to try_format may be added in the future that test other aspects, so this acts as a little delineator

l0rinc replied:
thanks for clarifying

stickies-v commented on src/netbase.cpp:
```diff
@@ -559,12 +559,8 @@ std::function<std::unique_ptr<Sock>(int, int, int)> CreateSock = CreateSockOS;
-    } else {
-        LogDebug(BCLog::NET, "%s\n", error_message);
-    }
+    auto log_level{manual_connection ? BCLog::Level::Info : BCLog::Level::Debug};
```
Thanks for pointing out the behaviour change, I've dropped the commit.

l0rinc reviewed (approved):
utACK 6dc6e4583c02f0d21dffeb6b2c81358b49a1f340

stickies-v commented on src/util/string.h:
```diff
@@ -235,11 +235,25 @@ template <typename T1, size_t PREFIX_LEN>
 namespace tinyformat {
 template <typename... Args>
-std::string format(util::ConstevalFormatString<sizeof...(Args)> fmt, const Args&... args)
+std::string try_format(util::ConstevalFormatString<sizeof...(Args)> fmt, const Args&... args)
```
I renamed it to `try_format` to be clear about the different behaviour compared to `tfm::format`, but agreed that avoiding merge conflicts is useful to coordinate on.

> My preference would be to leave the name as-is and instead just delete the unchecked one

Could you elaborate on this a bit more? We have these 4 other `format` overloads, I don't think you want to just remove them all? At least 2) and 4) would imply touching quite a bit of code to refactor, and it seems not all call sites of `format` are currently constant expressions?

1) `std::string format(const std::string &fmt, const Args&... args)`
2) `std::string format(const char* fmt, const Args&... args)`
3) `void format(std::ostream& out, const char* fmt, const Args&... args)`
4) `bilingual_str format(const bilingual_str& fmt, const Args&... args)`

stickies-v replied:
> so `format(const char*, ...)` would have to be deleted for the compiler to pick `format(ConstevalFormatString, ...)`

I think just deleting `format(const char*, ...)` leads to ambiguous overloading?

```
../../../../src/checkqueue.h:136:36: error: call to 'format' is ambiguous
                util::ThreadRename(strprintf("scriptch.%i", n));
                                   ^~~~~~~~~
../../../../src/tinyformat.h:1151:19: note: expanded from macro 'strprintf'
#define strprintf tfm::format
                  ^~~~~~~~~~~
../../../../src/tinyformat.h:1140:13: note: candidate function [with Args = <int>]
std::string format(const std::string &fmt, const Args&... args)
            ^
../../../../src/util/string.h:249:13: note: candidate function [with Args = <int>]
std::string format(util::ConstevalFormatString<sizeof...(Args)> fmt, const Args&... args)
```

> IIUC the remaining `format` taking a `std::string` can be left as-is for now

That (and keeping the `bilingual_str` overload) would be my preference too, but you also asked to remove `./test/lint/lint-format-strings.py` in this pull which I think we shouldn't do before all `tfm::format` overloads are compile-time checked? So that's why I'm a bit confused.

stickies-v replied:
Thanks for looking into this further! I was also exploring the concepts route, but hadn't yet resolved all concerns.

1) Re naming: having `format` overloads with significantly differently behaviour (throwing vs non-throwing) didn't quite sit right with me, so I think your `format_raw` (or `format_throw`) suggestion is preferable, and I agree it makes more sense to rename the throwing variant instead of the non-throwing variant (although the downside is that it makes reviewing this PR less straightforward).
2) forcing the `util::ConstevalFormatString` overload for all string literals either requires that:
    a) all callsites include `util/string.h`
    b) `tinyformat.h` includes `util/string.h` (as in your suggestion, although we should probably iwyu export it too), but this introduces a circular dependency
    c) the `util::ConstevalFormatString` overloads are moved to `tinyformat.h`, so the `tinyformat.h` dependency in `util/string.h` can be removed.

I think c) is the preferable approach. It adds more core-specific LoC to tinyformat but I think it is the cleanest overall?

> So an easy fix would be to just call the unchecked c-string functions in bitcoin-cli, which I think is fine.

I agree, although I think I'll clean up the incorrect `bitcoin-cli` format strings here too since that seems like an easy win.

stickies-v committed:
move-only: ConstevalFormatString tfm::format to tinyformat.h

In future commits, the non-ConstevalFormatString tfm::format
overloads will be renamed or removed. To avoid all callsites
having to include util/string.h (or to have a circular dependency
between tinyformat.h and util/string.h), just move the
ConstevalFormatString overloads to tinyformat.h

stickies-v committed:
tinyformat: force compile-time checks for format string literals

Callsites such as bitcoin-cli that (incorrectly) fail compile-time
validation can still use tfm::format_raw.

stickies-v committed:
tinyformat: remove std::string overload

All (non-test) format strings are known at compile-time, so we
don't need the std::string overload. Remove it so users use the
safer util::ConstevalFormatString where possible, or the
bilingual_str overload where applicable.

stickies-v committed:
tinyformat: don't throw format string errors

In almost all cases, invalid format strings should not lead to crashing
behaviour, so suppress run-time format errors by returning the error as
a string instead of throwing a tinyformat::format_error.

In DEBUG builds, tinyformat::format_error throwing behaviour is
preserved.

Note: most format string literals are partially validated at compile-time
through util::ConstevalFormatString. Notably, bilingual_str format strings
do not have this compile-time validation.

stickies-v commented:
Latest force-push increases the scope of this PR, see updated PR description for an overview. In a nutshell:
- all non-`bilingual_str` `format`/`strprintf` usage is now compile-time checked. Format errors are returned as strings instead of thrown as error, but throwing behaviour can be maintained through `tfm::format_raw`.
- the `const std::string&` `format` overload is removed because it's unnecessary

Thanks a lot @maflcko and @l0rinc for your review!

maflcko reviewed (approved):
lgtm

Thanks for taking the feedback

stickies-v commented on src/tinyformat.h:
```diff
@@ -1147,15 +1147,32 @@ TINYFORMAT_FOREACH_ARGNUM(TINYFORMAT_MAKE_FORMAT_FUNCS)
+    try {
+        tfm::format_raw(out, fmt.fmt, args...);
+    } catch (tinyformat::format_error& fmterr) {
+        out << "Error \"" + std::string{fmterr.what()} + "\" while formatting log message: \"" + fmt.fmt + "\"";
```
I assumed the next step would be to validate the `bilingual_str` overload at compile-time too, but I've been looking at implementing that and with the run-time translation dependency (`QCoreApplication::translate`) it doesn't really seem feasible.

I hadn't considered that translations can introduce format errors too, which makes the case for `bilingual_str` format errors not throwing ever stronger indeed. I was worried that blanket catching all formatting errors at the `tfm` level could be considered controversial so went for a throwing `format_raw` and non-throwing `format`, but perhaps just never having throwing behaviour makes for a more consistent and safer interface, so I'll move the try/catch to `tfm::vformat` now.

maflcko reviewed (approved):
lgtm

maflcko reviewed (approved):
review ACK bb12305737e3c22325930b5bdb04028f1a6f2ef6 💅

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: review ACK bb12305737e3c22325930b5bdb04028f1a6f2ef6 💅
Fge/JMP/cEBpCNfrVWYhr7mIRY4zXGiZRkBzWtbVouQ4q6WPnKlif9AdGqIb4Jrmkp/sIcvaECTG9QyCy1/mBA==
```

</details>

stickies-v commented on src/wallet/test/fuzz/notifications.cpp:
```diff
@@ -64,7 +64,7 @@ void ImportDescriptors(CWallet& wallet, const std::string& seed_insecure)
     for (const std::string& desc_fmt : DESCS) {
         for (bool internal : {true, false}) {
-            const auto descriptor{(strprintf)(desc_fmt.c_str(), "[5aa9973a/66h/4h/2h]" + seed_insecure, int{internal})};
+            const auto descriptor{(tfm::format_raw)(desc_fmt.c_str(), "[5aa9973a/66h/4h/2h]" + seed_insecure, int{internal})};
```
Ah, I wasn't sure why that was. Fuzzer compiles and linter seems happy so updated, thanks!

l0rinc commented on src/test/util_string_tests.cpp:
```diff
@@ -83,4 +86,15 @@ BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
     FailFmtWithError<1>("%1$", err_term);
 }
 
+BOOST_AUTO_TEST_CASE(tinyformat_ConstevalFormatString)
```
Unrelated:
just realized that we should eventually add `"%1"` passes, since `while ('0' <= *it && *it <= '9') {` can go beyond the string and passed accidentally - but seems to fail in tinyformat, see: https://godbolt.org/z/nbTKnrqch.
Should I add a new PR for that of do you want to add that case here?

stickies-v replied:
Looks like this is already covered in #30999, resolving this.

l0rinc replied:
Yes, your review is welcome there as well

l0rinc commented on src/bitcoin-cli.cpp:
```diff
@@ -540,15 +540,15 @@ class NetinfoRequestHandler : public BaseRequestHandler
-                                m_max_addr_processed_length, "addrp",
-                                m_max_addr_rate_limited_length, "addrl",
-                                m_max_age_length, "age");
+            result += tfm::format_raw("<->   type   net  v  mping   ping send recv  txn  blk  hb %*s%*s%*s ",
```
we could avoid formatting these by adding something like:
```C++
if (*it == '*' || ((*it == '_' || *it == '-') && *std::next(it) == '*')) {
    return; // We can't validate format string with %* or %_
}
```
to `Detail_CheckNumFormatSpecifiers` , which would enable us using `tfm::format` here - maybe later we'll implement full validation, but this way at least everything before the `*` format is validated properly.

l0rinc replied:
Weren't we already doing partial validation?
Doing a full one for these examples wouldn't be *that* difficult - do you suggest we do that instead?

l0rinc replied:
Alternatively, if that's not desired, what if we replaced the dynamic widths with explicit `PadLeft`/`PadRight` calls, e.g.
```C++
result += tfm::format("<->   type   net  v  mping   ping send recv  txn  blk  hb %s%s%s ",
                      PadRight("addrp", m_max_addr_processed_length),
                      PadRight("addrl", m_max_addr_rate_limited_length),
                      PadRight("age", m_max_age_length));
 ```
 and
 ```C++
 result += tfm::format("%s %s%s\n",
                      PadLeft("id", m_max_id_length),
                      IsAddressSelected() ? PadRight("address", m_max_addr_length) : "",
                      IsVersionSelected() ? "version" : "");
```
etc.
I can provide this in a follow-up, if the reviewers think it's a good direction.

<details>
<summary>PadLeft/PadRight</summary>

```C++
[[nodiscard]] inline std::string PadLeft(const std::string_view& str, size_t width, const char pad = ' ')
{
    if (str.size() >= width) return std::string(str);
    return std::string(width - str.size(), pad) + std::string(str);
}

[[nodiscard]] inline std::string PadRight(const std::string_view& str, size_t width, const char pad = ' ')
{
    if (str.size() >= width) return std::string(str);
    return std::string(str) + std::string(width - str.size(), pad);
}
```

</details>

l0rinc replied:
Extracted it to https://github.com/bitcoin/bitcoin/pull/30999/files#diff-71badc1cc71ba46244f7841a088251bb294265f4fe9e662c0ad6b15be540eee4R45

stickies-v replied:
Resolving since it has its own PR now.

l0rinc commented on src/test/util_string_tests.cpp:
```diff
@@ -83,4 +86,15 @@ BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
 
+BOOST_AUTO_TEST_CASE(tinyformat_ConstevalFormatString)
+{
+    // ensure invalid format strings don't throw at run-time
```
nit:
```suggestion
    // Ensure invalid format strings don't throw at run-time
```

l0rinc commented on src/test/util_string_tests.cpp:
```diff
@@ -83,4 +86,15 @@ BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
     FailFmtWithError<1>("%1$", err_term);
 }
 
```
We could add the examples from `bitcoin-cli` here:
```C++
BOOST_AUTO_TEST_CASE(ConstevalFormatString_SpecificUsageTests)
{
    // These format strings contain '*' and should skip validation
    PassFmt<6>("<->   type   net  v  mping   ping send recv  txn  blk  hb %*s%*s%*s ");
    PassFmt<5>("%*s %-*s%s\n");
    PassFmt<22>(
        "%3s %6s %5s %2s%7s%7s%5s%5s%5s%5s  %2s %*s%*s%*s%*i %*s %-*s%s\n");
    PassFmt<2>("                        ms     ms  sec  sec  min  min                %*s\n\n");
    PassFmt<4>("\n%-*s    port %6i    score %6i");
    PassFmt<3>("%*s %s\n");
}
```

stickies-v replied:
Addressed in #30999, resolving.

l0rinc reviewed (changes requested):
I have concerns about the translations, using raw format instead of adjusting the validator, and I think we could sneak in a fix for unterminated positionals.

<details>
<summary>Patch</summary>

```patch
Index: src/util/string.h
<+>UTF-8
===================================================================
diff --git a/src/util/string.h b/src/util/string.h
--- a/src/util/string.h	(revision 2d4f82d590845f9b2ea11bbcfe9377ef14df8775)
+++ b/src/util/string.h	(date 1727560531914)
@@ -51,12 +51,19 @@
                 ++it;
                 continue;
             }
+            if (*it == '*' || ((*it == '_' || *it == '-') && *std::next(it) == '*')) {
+                count_normal += 2; // width + value
+                it += 2;
+                continue;
+            }
 
             unsigned maybe_num{0};
             while ('0' <= *it && *it <= '9') {
                 maybe_num *= 10;
                 maybe_num += *it - '0';
-                ++it;
+                if (++it >= str.end()) throw "Format specifier incorrectly terminated by end of string";
             };
 
             if (*it == '$') {
Index: src/test/util_string_tests.cpp
<+>UTF-8
===================================================================
diff --git a/src/test/util_string_tests.cpp b/src/test/util_string_tests.cpp
--- a/src/test/util_string_tests.cpp	(revision 2d4f82d590845f9b2ea11bbcfe9377ef14df8775)
+++ b/src/test/util_string_tests.cpp	(date 1727560389361)
@@ -61,7 +61,6 @@
     // The `*` specifier behavior is unsupported and can lead to runtime
     // errors when used in a ConstevalFormatString. Please refer to the
     // note in the ConstevalFormatString docs.
-    PassFmt<1>("%*c");
     PassFmt<2>("%2$*3$d");
     PassFmt<1>("%.*f");
 
@@ -83,18 +82,27 @@
 
     auto err_term{"Format specifier incorrectly terminated by end of string"};
     FailFmtWithError<1>("%", err_term);
+    FailFmtWithError<1>("%1", err_term);
     FailFmtWithError<1>("%1$", err_term);
 }
 
+BOOST_AUTO_TEST_CASE(ConstevalFormatString_SpecificUsageTests)
+{
+    PassFmt<6>("<->   type   net  v  mping   ping send recv  txn  blk  hb %*s%*s%*s ");
+    PassFmt<5>("%*s %-*s%s\n");
+    PassFmt<24>( "%3s %6s %5s %2s%7s%7s%5s%5s%5s%5s  %2s %*s%*s%*s%*i %*s %-*s%s\n");
+    PassFmt<2>("                        ms     ms  sec  sec  min  min                %*s\n\n");
+    PassFmt<4>("\n%-*s    port %6i    score %6i");
+    PassFmt<3>("%*s %s\n");
+}
+
 BOOST_AUTO_TEST_CASE(tinyformat_ConstevalFormatString)
 {
-    // ensure invalid format strings don't throw at run-time
-    BOOST_CHECK_EQUAL(tfm::format("%*c", "dummy"), R"(Error "tinyformat: Cannot convert from argument type to integer for use as variable width or precision" while formatting log message: "%*c")");
+    // Ensure invalid format strings don't throw at run-time
     BOOST_CHECK_EQUAL(tfm::format("%2$*3$d", "dummy", "value"), R"(Error "tinyformat: Positional argument out of range" while formatting log message: "%2$*3$d")");
     std::ostringstream oss;
     tfm::format(oss, "%.*f", 5);
     BOOST_CHECK_EQUAL(oss.str(), R"(Error "tinyformat: Too many conversion specifiers in format string" while formatting log message: "%.*f")");
 }
 
-
 BOOST_AUTO_TEST_SUITE_END()
Index: src/bitcoin-cli.cpp
<+>UTF-8
===================================================================
diff --git a/src/bitcoin-cli.cpp b/src/bitcoin-cli.cpp
--- a/src/bitcoin-cli.cpp	(revision 2d4f82d590845f9b2ea11bbcfe9377ef14df8775)
+++ b/src/bitcoin-cli.cpp	(date 1727560250888)
@@ -540,42 +540,46 @@
         // Report detailed peer connections list sorted by direction and minimum ping time.
         if (DetailsRequested() && !m_peers.empty()) {
             std::sort(m_peers.begin(), m_peers.end());
-            result += tfm::format_raw("<->   type   net  v  mping   ping send recv  txn  blk  hb %*s%*s%*s ",
-                                      m_max_addr_processed_length, "addrp",
-                                      m_max_addr_rate_limited_length, "addrl",
-                                      m_max_age_length, "age");
+            result += tfm::format("<->   type   net  v  mping   ping send recv  txn  blk  hb %*s%*s%*s ",
+                                  m_max_addr_processed_length, "addrp",
+                                  m_max_addr_rate_limited_length, "addrl",
+                                  m_max_age_length, "age");
             if (m_is_asmap_on) result += " asmap ";
-            result += tfm::format_raw("%*s %-*s%s\n", m_max_id_length, "id", IsAddressSelected() ? m_max_addr_length : 0, IsAddressSelected() ? "address" : "", IsVersionSelected() ? "version" : "");
+            result += tfm::format("%*s %-*s%s\n",
+                                  m_max_id_length, "id",
+                                  IsAddressSelected() ? m_max_addr_length : 0,
+                                  IsAddressSelected() ? "address" : "",
+                                  IsVersionSelected() ? "version" : "");
             for (const Peer& peer : m_peers) {
                 std::string version{ToString(peer.version) + peer.sub_version};
-                result += tfm::format_raw(
-                    "%3s %6s %5s %2s%7s%7s%5s%5s%5s%5s  %2s %*s%*s%*s%*i %*s %-*s%s\n",
-                    peer.is_outbound ? "out" : "in",
-                    ConnectionTypeForNetinfo(peer.conn_type),
-                    peer.network,
-                    (peer.transport_protocol_type.size() == 2 && peer.transport_protocol_type[0] == 'v') ? peer.transport_protocol_type[1] : ' ',
-                    PingTimeToString(peer.min_ping),
-                    PingTimeToString(peer.ping),
-                    peer.last_send ? ToString(time_now - peer.last_send) : "",
-                    peer.last_recv ? ToString(time_now - peer.last_recv) : "",
-                    peer.last_trxn ? ToString((time_now - peer.last_trxn) / 60) : peer.is_tx_relay ? "" : "*",
-                    peer.last_blck ? ToString((time_now - peer.last_blck) / 60) : "",
-                    strprintf("%s%s", peer.is_bip152_hb_to ? "." : " ", peer.is_bip152_hb_from ? "*" : " "),
-                    m_max_addr_processed_length, // variable spacing
-                    peer.addr_processed ? ToString(peer.addr_processed) : peer.is_addr_relay_enabled ? "" : ".",
-                    m_max_addr_rate_limited_length, // variable spacing
-                    peer.addr_rate_limited ? ToString(peer.addr_rate_limited) : "",
-                    m_max_age_length, // variable spacing
-                    peer.age,
-                    m_is_asmap_on ? 7 : 0, // variable spacing
-                    m_is_asmap_on && peer.mapped_as ? ToString(peer.mapped_as) : "",
-                    m_max_id_length, // variable spacing
-                    peer.id,
-                    IsAddressSelected() ? m_max_addr_length : 0, // variable spacing
-                    IsAddressSelected() ? peer.addr : "",
-                    IsVersionSelected() && version != "0" ? version : "");
+                result += tfm::format("%3s %6s %5s %2s%7s%7s%5s%5s%5s%5s  %2s %*s%*s%*s%*i %*s %-*s%s\n",
+                                      peer.is_outbound ? "out" : "in",
+                                      ConnectionTypeForNetinfo(peer.conn_type),
+                                      peer.network,
+                                      (peer.transport_protocol_type.size() == 2 && peer.transport_protocol_type[0] == 'v') ? peer.transport_protocol_type[1] : ' ',
+                                      PingTimeToString(peer.min_ping),
+                                      PingTimeToString(peer.ping),
+                                      peer.last_send ? ToString(time_now - peer.last_send) : "",
+                                      peer.last_recv ? ToString(time_now - peer.last_recv) : "",
+                                      peer.last_trxn ? ToString((time_now - peer.last_trxn) / 60) : peer.is_tx_relay ? "" : "*",
+                                      peer.last_blck ? ToString((time_now - peer.last_blck) / 60) : "",
+                                      strprintf("%s%s", peer.is_bip152_hb_to ? "." : " ", peer.is_bip152_hb_from ? "*" : " "),
+                                      m_max_addr_processed_length, // variable spacing
+                                      peer.addr_processed ? ToString(peer.addr_processed) : peer.is_addr_relay_enabled ? "" : ".",
+                                      m_max_addr_rate_limited_length, // variable spacing
+                                      peer.addr_rate_limited ? ToString(peer.addr_rate_limited) : "",
+                                      m_max_age_length, // variable spacing
+                                      peer.age,
+                                      m_is_asmap_on ? 7 : 0, // variable spacing
+                                      m_is_asmap_on && peer.mapped_as ? ToString(peer.mapped_as) : "",
+                                      m_max_id_length, // variable spacing
+                                      peer.id,
+                                      IsAddressSelected() ? m_max_addr_length : 0, // variable spacing
+                                      IsAddressSelected() ? peer.addr : "",
+                                      IsVersionSelected() && version != "0" ? version : "");
             }
-            result += tfm::format_raw("                        ms     ms  sec  sec  min  min                %*s\n\n", m_max_age_length, "min");
+            result += tfm::format("                        ms     ms  sec  sec  min  min                %*s\n\n",
+                                  m_max_age_length, "min");
         }
 
         // Report peer connection totals by type.
@@ -624,7 +628,10 @@
                 max_addr_size = std::max(addr["address"].get_str().length() + 1, max_addr_size);
             }
             for (const UniValue& addr : local_addrs) {
-                result += tfm::format_raw("\n%-*s    port %6i    score %6i", max_addr_size, addr["address"].get_str(), addr["port"].getInt<int>(), addr["score"].getInt<int>());
+                result += tfm::format("\n%-*s    port %6i    score %6i",
+                                      max_addr_size, addr["address"].get_str(),
+                                      addr["port"].getInt<int>(),
+                                          addr["score"].getInt<int>());
             }
         }
 
@@ -1116,10 +1123,9 @@
         }
 
         for (const std::string& wallet : result["balances"].getKeys()) {
-            result_string += tfm::format_raw("%*s %s\n",
-                                             max_balance_length,
-                                             result["balances"][wallet].getValStr(),
-                                             wallet.empty() ? "\"\"" : wallet);
+            result_string += tfm::format("%*s %s\n",
+                                         max_balance_length, result["balances"][wallet].getValStr(),
+                                         wallet.empty() ? "\"\"" : wallet);
         }
         result_string += "\n";
     }
Index: src/test/txrequest_tests.cpp
<+>UTF-8
===================================================================
diff --git a/src/test/txrequest_tests.cpp b/src/test/txrequest_tests.cpp
--- a/src/test/txrequest_tests.cpp	(revision 2d4f82d590845f9b2ea11bbcfe9377ef14df8775)
+++ b/src/test/txrequest_tests.cpp	(date 1727560250869)
@@ -181,7 +181,7 @@
             BOOST_CHECK_MESSAGE(real_total == total, strprintf("[%s] total %i (%i expected)", comment, real_total, total));
             BOOST_CHECK_MESSAGE(real_inflight == inflight, strprintf("[%s] inflight %i (%i expected)", comment, real_inflight, inflight));
             BOOST_CHECK_MESSAGE(real_candidates == candidates, strprintf("[%s] candidates %i (%i expected)", comment, real_candidates, candidates));
-            BOOST_CHECK_MESSAGE(ret == expected, "[" + comment + "] mismatching requestables");
+            BOOST_CHECK_MESSAGE(ret == expected, strprintf("[%s] mismatching requestables", comment));
         });
     }
 ```

</details>

stickies-v commented on src/clientversion.cpp:
```diff
@@ -87,15 +87,17 @@ std::string LicenseInfo()
     const std::string URL_SOURCE_CODE = "<https://github.com/bitcoin/bitcoin>";
 
-    return CopyrightHolders(strprintf(_("Copyright (C) %i-%i").translated, 2009, COPYRIGHT_YEAR) + " ") + "\n" +
+    return CopyrightHolders(strprintf(_("Copyright (C) %i-%i"), 2009, COPYRIGHT_YEAR).translated + " ") + "\n" +
```
`"Copyright (C) %i-%i"` remains the translateable string both before and after this change. The only difference is that we use the `bilingual_str` `format` overload instead of the `const std::string&` one.

l0rinc replied:
Got it, thanks for clarifying!

l0rinc commented on src/bitcoin-cli.cpp:
```diff
@@ -540,15 +540,15 @@ class NetinfoRequestHandler : public BaseRequestHandler
             for (const Peer& peer : m_peers) {
                 std::string version{ToString(peer.version) + peer.sub_version};
-                result += strprintf(
+                result += tfm::format_raw(
```
do we need to keep the `strprintf("%s%s"` on line 563?

stickies-v replied:
I suppose we don't, but that seems unrelated to this PR so going to leave that as is.

l0rinc replied:
Can you expand on that? We've changed other `strprintf` instance here to enable compile-time validation, how come this one didn't make the cut?

stickies-v replied:
Hmm, then I don't think I understand what you mean. Would you be able to provide a diff/example?

l0rinc replied:
My mistake, rereading it it's not clear what I was asking.
So I think we could simplify the nested formats by pulling up the internal pattern:
Having `%2s` in the main pattern and having a `%s%s` inside could be simplified to having `%s%s` at the parent and layout out the values in the parameters:
```patch
Index: src/bitcoin-cli.cpp
<+>UTF-8
===================================================================
diff --git a/src/bitcoin-cli.cpp b/src/bitcoin-cli.cpp
--- a/src/bitcoin-cli.cpp	(revision 2e33dbadcf69838c8ea591dccdc666f809c3134a)
+++ b/src/bitcoin-cli.cpp	(date 1727693490146)
@@ -549,7 +549,7 @@
             for (const Peer& peer : m_peers) {
                 std::string version{ToString(peer.version) + peer.sub_version};
                 result += tfm::format_raw(
-                    "%3s %6s %5s %2s%7s%7s%5s%5s%5s%5s  %2s %*s%*s%*s%*i %*s %-*s%s\n",
+                    "%3s %6s %5s %2s%7s%7s%5s%5s%5s%5s  %s%s %*s%*s%*s%*i %*s %-*s%s\n",
                     peer.is_outbound ? "out" : "in",
                     ConnectionTypeForNetinfo(peer.conn_type),
                     peer.network,
@@ -560,7 +560,8 @@
                     peer.last_recv ? ToString(time_now - peer.last_recv) : "",
                     peer.last_trxn ? ToString((time_now - peer.last_trxn) / 60) : peer.is_tx_relay ? "" : "*",
                     peer.last_blck ? ToString((time_now - peer.last_blck) / 60) : "",
-                    strprintf("%s%s", peer.is_bip152_hb_to ? "." : " ", peer.is_bip152_hb_from ? "*" : " "),
+                    peer.is_bip152_hb_to ? "." : " ",
+                    peer.is_bip152_hb_from ? "*" : " ",
                     m_max_addr_processed_length, // variable spacing
                     peer.addr_processed ? ToString(peer.addr_processed) : peer.is_addr_relay_enabled ? "" : ".",
                     m_max_addr_rate_limited_length, // variable spacing
Index: src/test/util_string_tests.cpp
<+>UTF-8
===================================================================
diff --git a/src/test/util_string_tests.cpp b/src/test/util_string_tests.cpp
--- a/src/test/util_string_tests.cpp	(revision 2e33dbadcf69838c8ea591dccdc666f809c3134a)
+++ b/src/test/util_string_tests.cpp	(date 1727693570815)
@@ -96,5 +96,10 @@
     BOOST_CHECK_EQUAL(oss.str(), R"(Error "tinyformat: Too many conversion specifiers in format string" while formatting log message: "%.*f")");
 }
 
+BOOST_AUTO_TEST_CASE(TwoCharFormattingTest)
+{
+    BOOST_CHECK_EQUAL(tfm::format("%2s", strprintf("%s%s", ".", "*")),
+                      tfm::format("%s%s", ".", "*"));
+}
 
 BOOST_AUTO_TEST_SUITE_END()
```

(the dummy test was added just to make sure we're doing it correctly)

stickies-v replied:
Thanks for providing the diff. There are probably plenty of opportunities to clean up the bitcoin-cli string printing, but this line is not touched and it does not relate to the goal of this PR, so it seems best to me to leave this as is.

l0rinc replied:
Ok, thanks for checking

stickies-v commented on src/test/txrequest_tests.cpp:
```diff
@@ -178,9 +178,9 @@ class Scenario
+            BOOST_CHECK_MESSAGE(real_total == total, strprintf("[%s] total %i (%i expected)", comment, real_total, total));
+            BOOST_CHECK_MESSAGE(real_inflight == inflight, strprintf("[%s] inflight %i (%i expected)", comment, real_inflight, inflight));
+            BOOST_CHECK_MESSAGE(real_candidates == candidates, strprintf("[%s] candidates %i (%i expected)", comment, real_candidates, candidates));
             BOOST_CHECK_MESSAGE(ret == expected, "[" + comment + "] mismatching requestables");
```
I think `BOOST_TEST_CONTEXT` would simplify this whole block quite nicely, but it's out of scope for this PR so I'm going not going to push that here. I'll update l184 with a `strprintf` as per your suggestion.

<details>
<summary>git diff on 2d4f82d590</summary>

```diff
diff --git a/src/test/txrequest_tests.cpp b/src/test/txrequest_tests.cpp
index a6c4a3613d..e70d9aa826 100644
--- a/src/test/txrequest_tests.cpp
+++ b/src/test/txrequest_tests.cpp
@@ -164,7 +164,6 @@ public:
         size_t completed, const std::string& checkname,
         std::chrono::microseconds offset = std::chrono::microseconds{0})
     {
-        const auto comment = m_testname + " " + checkname;
         auto& runner = m_runner;
         const auto now = m_now;
         assert(offset.count() <= 0);
@@ -178,10 +177,12 @@ public:
             size_t real_total = runner.txrequest.Count(peer);
             size_t real_candidates = runner.txrequest.CountCandidates(peer);
             size_t real_inflight = runner.txrequest.CountInFlight(peer);
-            BOOST_CHECK_MESSAGE(real_total == total, strprintf("[%s] total %i (%i expected)", comment, real_total, total));
-            BOOST_CHECK_MESSAGE(real_inflight == inflight, strprintf("[%s] inflight %i (%i expected)", comment, real_inflight, inflight));
-            BOOST_CHECK_MESSAGE(real_candidates == candidates, strprintf("[%s] candidates %i (%i expected)", comment, real_candidates, candidates));
-            BOOST_CHECK_MESSAGE(ret == expected, "[" + comment + "] mismatching requestables");
+            BOOST_TEST_CONTEXT(m_testname + " " + checkname) {
+                BOOST_CHECK_EQUAL(real_total, total);
+                BOOST_CHECK_EQUAL(real_inflight, inflight);
+                BOOST_CHECK_EQUAL(real_candidates, candidates);
+                BOOST_CHECK_EQUAL_COLLECTIONS(ret.begin(), red.end(), expected.begin(), expected.end());
+            }
         });
     }
 

```
</details>

l0rinc replied:
wasn't aware of `BOOST_TEST_CONTEXT`, would indeed be an improvement here

stickies-v commented on doc/developer-notes.md:
```diff
@@ -950,7 +950,9 @@ Strings and formatting
   - Do not use it when passing strings to `tfm::format`, `strprintf`, `LogInfo`, `LogDebug`, etc.
 
-    - *Rationale*: This is redundant. Tinyformat handles strings.
+    - *Rationale*: Tinyformat handles string parameters. Format strings
+      should be known at compile-time, so using a string literal or
+      `constexpr const char*` allows for compile-time validation.
```
I don't think this is an improvement, I find it less clear about the distinction between the compile-time format string and the run-time string parameters, so I'm gong to leave this as is for now.

l0rinc replied:
Not sure why it's less clear than before, but it's just a nit from my part anyway

stickies-v replied:
I used `std::string` at first too, but then it was inconsistent with the rest of the documentation, which I think is even more unclear. I'll reconsider updating all `std::string` references in this block if I have to force-push again unless you think this is blocking?

hodlinator replied:
Changing tinyformat to support `string_view` is definitely out of scope for this PR.

I would at least want to make it clearer that format strings are separate from other parameters, like this or with other words to that effect:
```suggestion
    - *Rationale*: Tinyformat handles string parameters. Format strings
      on the other hand should be known at compile-time, so using a string literal or
      `constexpr const char*` allows for compile-time validation.
```
Even though I decided to drop into l0rinc's thread here, I independently found it unclear when doing a review-pass on it.

More precise suggestion, but not blocking:
```suggestion
    - *Rationale*: Tinyformat handles string parameters. Format strings
      on the other hand should (except for translations) be known at compile-time, so using a string literal or
      `constexpr const char*` allows for compile-time validation.
```

stickies-v replied:
> I would at least want to make it clearer that format strings are separate from other parameters

I agree with that, and I'll incorporate your first suggestion (I [don't agree](https://github.com/bitcoin/bitcoin/pull/30928#issuecomment-2388899694) with the translations bit) if I have to force-push, thanks!

stickies-v commented on src/test/fuzz/strprintf.cpp:
```diff
@@ -48,35 +48,32 @@ FUZZ_TARGET(str_printf)
     //
     // Upstream bug report: https://github.com/c42f/tinyformat/issues/70
```
That seems unrelated to this PR though?

l0rinc replied:
The comment states:
```
    // Avoid triggering the following crash bug:
    // * strprintf("%.1s", (char*)nullptr);
```
but we're not calling `strprintf` directly anymore, hence my comment

l0rinc replied:
Ok, it's not that important

stickies-v commented:
Force-pushed to address two nits from @l0rinc - thanks for the review!

l0rinc commented:
Most of my observations should rather be addressed in a [separate PR](https://github.com/bitcoin/bitcoin/pull/30999) (can be merged before or after, I don't mind rebasing, would appreciate a review there).
I'm fine with merging this one as is, thanks for the improvement.

ACK 2e33dbadcf69838c8ea591dccdc666f809c3134a

l0rinc commented on src/test/util_string_tests.cpp:
```diff
@@ -83,4 +86,15 @@ BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
+BOOST_AUTO_TEST_CASE(tinyformat_ConstevalFormatString)
+{
+    // Ensure invalid format strings don't throw at run-time.
+    BOOST_CHECK_EQUAL(tfm::format("%*c", "dummy"), R"(Error "tinyformat: Cannot convert from argument type to integer for use as variable width or precision" while formatting log message: "%*c")");
```
you should probably update these tests as well

stickies-v commented on src/tinyformat.h:
```diff
@@ -1049,7 +1049,12 @@ TINYFORMAT_FOREACH_ARGNUM(TINYFORMAT_MAKE_MAKEFORMATLIST)
+    try {
+        detail::formatImpl(out, fmt, list.m_args, list.m_N);
+    } catch (tinyformat::format_error& fmterr) {
+        out << "Error \"" + std::string{fmterr.what()} + "\" while formatting log message: \"" + fmt + "\"";
```
Ah, good catch, thanks. Updated to remove the quotes and revert to existing behaviour, to avoid messing up the logs like e.g. the below:

With quotes:
```
2024-09-30T12:22:36Z Warning: Error "tinyformat: Cannot convert from argument type to integer for use as variable width or precision" while formatting log message: "Warning: %*s Support for testnet3 is deprecated and will be removed in an upcoming release. Consider switching to testnet4.
"Script verification uses 7 additional threads
```
Without quotes:
```
2024-09-30T12:21:18Z Warning: Error "tinyformat: Cannot convert from argument type to integer for use as variable width or precision" while formatting log message: Warning: %*s Support for testnet3 is deprecated and will be removed in an upcoming release. Consider switching to testnet4.
2024-09-30T12:21:18Z Script verification uses 7 additional threads
```

l0rinc commented:
ACK 4f33e9a85c3a8f546deddc2f78cf74bceff30315

hodlinator reviewed (commented):
Concept ACK 4f33e9a85c3a8f546deddc2f78cf74bceff30315

It would be cleaner *in the implementation* to use `util::Result` instead of overwriting the returned value with the error message, but it would mess up the call sites too much probably. Current approach in this PR seems one of the least worse ones.

Passed locally: `FUZZ=str_printf src/test/fuzz/fuzz ../qa-assets/fuzz_seed_corpus/str_printf/*`.

### PR summary nit
> rename the throwing `tfm::format` functions to `tfm::format_raw`

The last commit makes it so that none of the `format_raw` functions throw either?
`format_raw`ends up meaning "call with raw `const char* fmt`, not `ConstevalFormatString<> fmt`".

stickies-v commented:
> It would be cleaner _in the implementation_ to use `util::Result` instead of overwriting the returned value with the error message

I don't think that's a good approach for our codebase, even if there was minimal code churn. All format strings are known at compile-time, so any formatting errors are programming errors. Handling that with `util::Result` would be both a cumbersome interface and bad practice imo. We don't need to let callsites handle formatting errors, they should never occur, but if/when they do after all we don't want them to crash the application because they are harmless to program execution, until we fix the error in the code.

> The last commit makes it so that none of the `format_raw` functions throw either?

Sorry, missed that line when updating the PR description, fixed now!

stickies-v commented:
> In the PR description: I think this should also say `or bitcoin-cli`.

`bitcoin-cli` wasn't using the `const std::string&` overload which that bulletpoint is about. It is mentioned in the next bulletpoint about compile-time checks:

> Callsites that for some reason don't pass the compile-time checks (such as in `bitcoin-cli.cpp`) can use `tfm::format_raw`.

hodlinator reviewed (commented):
stickies-v:
> All format strings are known at compile-time, so any formatting errors are programming errors. Handling that with `util::Result` would be both a cumbersome interface and bad practice imo.

This is a good argument except we do have some strings only known at runtime, notably translations, which probably don't get full testing. Maybe there's something that automatically compares format strings for equivalence across translations though.

maflcko:
> I agree that `util::Result` doesn't make any sense here.

(See above. I'd say it makes *some* sense).

> Apart from the reasons mentioned, it would also make it harder to switch to `std::format`, as well as being confusing, because no other format lib I am aware of is doing that.

"errors as values" rather than errors as exceptions are generally a step in the right direction for runtime errors. Modern formatting libraries tend to use a lot of compile time checking though AFAIK, decreasing the need for runtime errors, so I'm happy about that direction in Bitcoin Core.

One could argue that if we returned `util::Result`, most call sites would probably opt to print an error + the offending format string to the log anyway, so doing that as a blanket rule for everything is acceptable.

stickies-v commented:
> This is a good argument except we do have some strings only known at runtime, notably translations

I chose my wording "known at compile-time" carefully for exactly this reason: they're _known_ at compile-time, we just choose not to implement the (complete) validation logic because it much more complex than just returning an error string. It's not worth arguing semantics too much, but in my view that's very much still a programming error and not something that callsites should be aware of, let alone handle.

ryanofsky commented:
Will not nack this pr since I don't think consequences are great, but I would prefer not to see this PR merged.

I don't think putting "error while formatting log message" in a std::string and then returning that std::string is a good way for a C++ API to return errors, especially errors that are almost certainly bugs in the code. Throwing an exception seems far more appropriate, and more likely to result in bugs being detected and fixed. (Also nit: the wording of that message doesn't make sense because tinyformat is used to format things other than log messages, and is also useful for low level string manipulation and things like formatting numbers or constructing urls)

I understand that practically speaking impact of this change should be small because the new compile time checking is pretty good, but that just means to my mind that there should be less of a need for this PR.

hodlinator commented:
My [Concept ACK](https://github.com/bitcoin/bitcoin/pull/30928#pullrequestreview-2339765974) accepted the current concept, while stating that I see it as one of the *least* worst alternatives (implying I don't see the error handling part as blocking). I'm happy to get feedback on that but my intent was not to start a debate.

@ryanofsky how would you feel about adding an `Assume()` statement that could fail Debug-builds when formatting fails?

ryanofsky commented:
Thanks, I didn't read the previous discussion history, so my feedback is only based on the implementation of this PR. not on earlier comments.

I think tinyformat's original decision to throw exceptions when invalid format strings were specified was flexible and sound, and the additional consteval checking we were able to add on top of that was a nice improvement that should be extended more places.

I just don't see a need to rip out tinyformat's simple handling mechanism and replace it with more complicated and unusual one, whether it's a self-referencing string apologizing for not being formatted correctly, or an inconsistently enforced check that behaves differently based on build flags.

To put it another way:

- I think adding consteval format string checking was a great improvement that should improve developer experience and catch bugs.
- I think the consteval string checking should be extended more places, for example supporting the `_` function to check translated original strings, and maybe even the imported translations of those strings.
- I don't like the runtime behavior change implemented in the fourth commit of this PR, and I'm not sure about the other commits. The only cited benefit in the PR description is "Avoid unexpected run-time crashes from string formatting" but I feel like the way to achieve that is to extend compile time checks more places, not to take a detour into changing runtime behavior.

hodlinator commented:
> My [Concept ACK](https://github.com/bitcoin/bitcoin/pull/30928#pullrequestreview-2339765974) accepted the current concept, while stating that I see it as one of the _least_ worst alternatives (implying I don't see the error handling part as blocking). I'm happy to get feedback on that but my intent was not to start a debate.

(This part my previous comment was more for maflcko & stickies-v. I think they were interpreting me as being more critical of the error-handling than I intended).

That said, I also see ryanofsky's perspective. The PR is borderline over-incremental instead of extending build time checks.

Worst case example that could help motivate the current PR approach:

A formatting error in a rarely used codepath in the Icelandic translation causes a unhandled exception, leading a lightning node to becoming unable to broadcast a penalty transaction in time. => Angry angry vikings. :crossed_swords:

stickies-v commented:
Force-pushed to rebase addressing a merge conflict, preserve `tfm::format_error` throwing behaviour for `DEBUG` builds to improve visibility into programming errors before they manifest, and [improve an error string](https://github.com/bitcoin/bitcoin/pull/30928#issuecomment-2389355374).

---

Thank you for your thoughtful review @ryanofsky. tl;dr I share your concerns, but I think I have a different opinion on the trade-offs made.

I think everyone here is in agreement that compile-time checks (ideally with `std::format`) are the proper way forward, and everything else is a temporary patch. I think there are differing views on what we do in the meanwhile. It appears, especially without `std::format` and having to deal with unverified translated strings, that implementing bulletproof compile-time checking on format strings is not really feasible or the best use of our time.

> I don't think putting "error while formatting log message" in a std::string and then returning that std::string is a good way for a C++ API to return errors, especially errors that are almost certainly bugs in the code.

I agree that returning error strings is not a good way to handle errors in general, but I don't think throwing a `std::runtime_error` is a _good_ way to deal with _programming_ errors either. Typically, these would be handled with `assert` statements (which is also `tinyformat`'s default), but crashing the node for malformed (usually) informational strings seems perhaps unnecessarily strict, in similar vein to our usage of `CHECK_NONFATAL` instead of `assert`?

Can you agree that the concept of not exposing `tfm::format_error` to callsites makes sense in our codebase, at least philosophically? Fomat errors _should_ never occur at run-time, even if they still can at present. If you agree with that, then I think it all comes down to what we do before we have full compile-time checking? It appears you prefer keeping the increased visibility when errors occur, whereas in this PR I prioritize not crashing the software in case we do miss it or if it's part of an erroneously translated format string.

I've force-pushed to preserve `tfm::format_error` throwing behaviour in `DEBUG` builds to keep more of the visibility without reducing stability. If you still feel like that new behaviour is still not an improvement over `master` (which is a view I can absolutely understand), I will drop the last commit and rework the first ones a bit. Even though the intent of the original version of this PR was to fix a [regression](https://github.com/bitcoin/bitcoin/pull/30889#discussion_r1766633746), I'd be happy to change course and keep just the first commits that force more compile-time checks, without changing run-time behaviour, which I think everyone here agrees is an improvement.

> EDIT: I also think the new uses of .c_str() are gross, though I didn't look into why those are necessary and that is not my main objection to this PR.

The `const std::string&` overload was dropped so that the `util::ConstevalFormatString` overload could be forced for all string literals. Although I don't like the additional `.c_str()` usage this introduced, it is mostly limited to fuzz and bitcoin-cli, and I think is worth the trade-off?

> I think tinyformat's original decision to throw exceptions when invalid format strings were specified was flexible and sound

Note: `tinyformat` natively `assert`s that there are no formatting errors, the `tfm::format_error` is Bitcoin Core-specific.

> or an inconsistently enforced check that behaves differently based on build flags.

I don't think this is correct: https://github.com/bitcoin/bitcoin/blob/caf44e500eb257376c36c0e043f71f72b949f743/src/tinyformat.h#L139

> (Also nit: the wording of that message doesn't make sense because tinyformat is used to format things other than log messages, and is also useful for low level string manipulation and things like formatting numbers or constructing urls)

Oops, good catch, thanks. This string was lifted from `logging.h` and I didn't properly update it - done now.

> how would you feel about adding an `Assume()` statement that could fail Debug-builds when formatting fails?

Adding `util/check.h` to `tinyformat.h` introduced a couple of circular dependencies so instead I'm using `#ifdef DEBUG`, but the effect should be the same! (We _could_ expand on this adding a `TFM_THROW_FORMAT_ERROR` flag (that defaults to set in `DEBUG`) so it can be added to more CI jobs without overhead if people would like that).

maflcko reviewed (commented):
Left a nit.

Only change is adding `throw` when `DEBUG`, which seems fine. Edit: However, I think any attempt to catch bugs at runtime in this context is fragile and inefficient. If bugs should be found, it would be better to catch them at (or before) compile-time with the compiler, or with a tool.

re-ACK 49f73071a8ec4131595090a619d6198a5f8b7c3c 🎖

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: re-ACK 49f73071a8ec4131595090a619d6198a5f8b7c3c 🎖
XSyui/vuov5yPzjlUZ4UeGQZHo/JZ6LsYy2ftwOqSStKhn/L2aHSd/uDBco6LvvTWd3tPsTlSY2y9myvd7H5Aw==
```

</details>

hodlinator commented:
Worth bringing back the `try`/`catch` in *src/test/fuzz/strprintf.cpp* for Debug builds?
```
₿ cmake --preset=libfuzzer -DCMAKE_BUILD_TYPE=Debug
...
₿ cmake --build build_fuzz -j<X>
...
₿ FUZZ=str_printf build_fuzz/src/test/fuzz/fuzz ../qa-assets/fuzz_seed_corpus/str_printf/
INFO: Running with entropic power schedule (0xFF, 100).
INFO: Seed: 734467281
INFO: Loaded 1 modules   (1292656 inline 8-bit counters): 1292656 [0x5d5503ee60a0, 0x5d5504021a10), 
INFO: Loaded 1 PC tables (1292656 PCs): 1292656 [0x5d5504021a10,0x5d55053db110), 
INFO:     1133 files found in ../qa-assets/fuzz_seed_corpus/str_printf/
INFO: -max_len is not provided; libFuzzer will not generate inputs larger than 16446 bytes
terminate called after throwing an instance of 'tinyformat::format_error'
  what():  tinyformat: Not enough conversion specifiers in format string
==59713== ERROR: libFuzzer: deadly signal
    #0 0x5d54f8395fea  (/home/hodlinator/bitcoin/build_fuzz/src/test/fuzz/fuzz+0x5bd1fea)
...
    #30 0x5d54f8245174  (/home/hodlinator/bitcoin/build_fuzz/src/test/fuzz/fuzz+0x5a81174)

NOTE: libFuzzer has rudimentary signal handlers.
      Combine libFuzzer with AddressSanitizer or similar for better crash reports.
SUMMARY: libFuzzer: deadly signal
MS: 0 ; base unit: 0000000000000000000000000000000000000000


artifact_prefix='./'; Test unit written to ./crash-da39a3ee5e6b4b0d3255bfef95601890afd80709
Base64: 
```

l0rinc commented:
I'm not in love with the `DEBUG` separation, but I'm fine with it if others are.

ACK 49f73071a8ec4131595090a619d6198a5f8b7c3c

`git range-diff 4f33e9a85c3a8f546deddc2f78cf74bceff30315~4..4f33e9a85c3a8f546deddc2f78cf74bceff30315 aba9ce0eb6e67945f2209a17676f356fe9748
7e6..928538e3782ccbb22f54f4a3b1152b1faba95471`

<details>
<summary>Diff</summary>

```patch
1:  1f2d0ffd56 < -:  ---------- move-only: ConstevalFormatString tfm::format to tinyformat.h
2:  c937ebf1ce = 1:  efa6f3f372 tinyformat: remove std::string overload
3:  be7de807c8 = 2:  9329e334eb tinyformat: force compile-time checks for format string literals
4:  4f33e9a85c ! 3:  928538e378 tinyformat: don't throw format string errors
    @@ Commit message
         behaviour, so suppress run-time format errors by returning the error as
         a string instead of throwing a tinyformat::format_error.
     
    +    In DEBUG builds, tinyformat::format_error throwing behaviour is
    +    preserved.
    +
         Note: most format string literals are partially validated at compile-time
         through util::ConstevalFormatString. Notably, bilingual_str format strings
         do not have this compile-time validation.
    @@ src/logging.h: template <typename... Args>
     -        try {
     -            log_msg = tfm::format(fmt, args...);
     -        } catch (tinyformat::format_error& fmterr) {
    --            /* Original format string will have newline so don't add one here */
     -            log_msg = "Error \"" + std::string{fmterr.what()} + "\" while formatting log message: " + fmt.fmt;
     -        }
     +        const auto log_msg{tfm::format(fmt, args...)};
    @@ src/test/util_string_tests.cpp
      #include <util/string.h>
      
      #include <boost/test/unit_test.hpp>
    + #include <test/util/setup_common.h>
      
     +#include <sstream>
     +
    @@ src/test/util_string_tests.cpp: BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSp
      
     +BOOST_AUTO_TEST_CASE(tinyformat_ConstevalFormatString)
     +{
    -+    // Ensure invalid format strings don't throw at run-time.
    -+    BOOST_CHECK_EQUAL(tfm::format("%*c", "dummy"), R"(Error "tinyformat: Cannot convert from argument type to integer for use as variable width or precision" while formatting log message: %*c)");
    -+    BOOST_CHECK_EQUAL(tfm::format("%2$*3$d", "dummy", "value"), R"(Error "tinyformat: Positional argument out of range" while formatting log message: %2$*3$d)");
    ++    // Ensure invalid format strings don't throw at run-time, when not in DEBUG
    ++#ifndef DEBUG
    ++    BOOST_CHECK_EQUAL(tfm::format("%*c", "dummy"), R"(Error "tinyformat: Cannot convert from argument type to integer for use as variable width or precision" while formatting: %*c)");
    ++    BOOST_CHECK_EQUAL(tfm::format("%2$*3$d", "dummy", "value"), R"(Error "tinyformat: Positional argument out of range" while formatting: %2$*3$d)");
     +    std::ostringstream oss;
     +    tfm::format(oss, "%.*f", 5);
    -+    BOOST_CHECK_EQUAL(oss.str(), R"(Error "tinyformat: Too many conversion specifiers in format string" while formatting log message: %.*f)");
    ++    BOOST_CHECK_EQUAL(oss.str(), R"(Error "tinyformat: Too many conversion specifiers in format string" while formatting: %.*f)");
    ++#endif
     +}
     +
     +
    @@ src/tinyformat.h: TINYFORMAT_FOREACH_ARGNUM(TINYFORMAT_MAKE_MAKEFORMATLIST)
     +    try {
     +        detail::formatImpl(out, fmt, list.m_args, list.m_N);
     +    } catch (tinyformat::format_error& fmterr) {
    -+        out << "Error \"" + std::string{fmterr.what()} + "\" while formatting log message: " + fmt;
    ++        out << "Error \"" + std::string{fmterr.what()} + "\" while formatting: " + fmt;
    ++#ifdef DEBUG
    ++        throw;
    ++#endif
     +    }
      }
```

</details>

> Worth bringing back the try/catch in src/test/fuzz/strprintf.cpp for Debug builds?

We should definitely be able to see easily what the invalid values were (and add unit tests for those cases)

stickies-v commented:
I've carved out the first 3 commits into https://github.com/bitcoin/bitcoin/pull/31149, since other PRs (e.g. #31061 and #31072) somewhat rely on this, and everyone seems in agreement that adding compile-time checks is good. Existing ACKs should be trivially transferable.

Converting this PR to draft until #31149 is merged to keep the controversial error suppression discussion until later.

hodlinator commented:
> > Worth bringing back the try/catch in src/test/fuzz/strprintf.cpp for Debug builds?
> 
> We should definitely be able to see easily what the invalid values were (and add unit tests for those cases)

(Not sure that would add much value, we know the fuzzing will produce mostly garbage format strings that will trigger errors).

stickies-v commented:
Closing this PR. Thanks to related PRs like #31174, #31072 and #31061, tfm::format error throwing should be significantly reduced, so there's increasingly little point in making this behaviour change.
